File: leetcode/problems/cpp/binary-tree-inorder-traversal.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(n)

#include "vector"
using namespace std;

struct TreeNode {
   int val;
   TreeNode* left;
   TreeNode* right;
   TreeNode() : val(0), left(nullptr), right(nullptr) {}
   TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
   TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
};

class Solution {
   vector<int> result;
   void inorder(TreeNode* node) {
      if (!node) return;
      inorder(node->left);
      result.push_back(node->val);
      inorder(node->right);
   }
public:
   vector<int> inorderTraversal(TreeNode* root) {
      inorder(root);
      return result;
   }
};

File: leetcode/problems/rs/binary-tree-inorder-traversal.rs
// by @codeAbinash
// Time : O(n)
// Space : O(n)

impl Solution {
    fn inorder(node: &Option<Rc<RefCell<TreeNode>>>, result: &mut Vec<i32>) {
        if let Some(v) = node {
            let v = v.borrow();
            Solution::inorder(&v.left, result);
            result.push(v.val);
            Solution::inorder(&v.right, result);
        }
    }
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut result = vec![];
        if let Some(v) = root {
            Solution::inorder(&Some(v), &mut result);
        }
        result
    }
}
